digdag-standards/src/main/java/io/digdag/standards/operator/aws/EmrOperatorFactory.java

```diff
@@ -0,0 +1,1753 @@
+                    .filename("config.json")
+                    .build();
+            RemoteFile remoteConfigurationFile = filer.prepareRemoteFile(tag, "bootstrap", Integer.toString(index), configurationFileReference, false);
+
+            return new BootstrapActionConfig()
+                    .withName(name)
+                    .withScriptBootstrapAction(new ScriptBootstrapActionConfig()
+                            .withPath(runner.s3Uri().toString())
+                            .withArgs(remoteConfigurationFile.s3Uri().toString()));
+        }
+
+        private String bootstrapWorkingDirectory(int index)
+        {
+            return LOCAL_STAGING_DIR + "/bootstrap/" + index + "/wd";
+        }
+
+        private List<Configuration> configurations(JsonNode node)
+        {
+            if (node.isTextual()) {
+                // File
+                String configurationJson;
+                try {
+                    configurationJson = workspace.templateFile(templateEngine, node.asText(), UTF_8, params);
+                }
+                catch (IOException | TemplateException e) {
+                    throw new TaskExecutionException(e, TaskExecutionException.buildExceptionErrorConfig(e));
+                }
+                List<ConfigurationJson> values;
+                try {
+                    values = objectMapper.readValue(configurationJson, new TypeReference<List<ConfigurationJson>>() {});
+                }
+                catch (IOException e) {
+                    throw new ConfigException("Invalid EMR configuration file: " + node.asText());
+                }
+                return values.stream()
+                        .map(ConfigurationJson::toConfiguration)
+                        .collect(toList());
+            }
+            else if (node.isObject()) {
+                // Embedded configuration
+                Config config = cf.create(node);
+                return ImmutableList.of(new Configuration()
+                        .withConfigurations(config.getListOrEmpty("configurations", JsonNode.class).stream()
+                                .map(this::configurations)
+                                .flatMap(Collection::stream)
+                                .collect(toList()))
+                        .withClassification(config.get("classification", String.class, null))
+                        .withProperties(config.get("properties", new TypeReference<Map<String, String>>() {}, null)));
+            }
+            else {
+                throw new ConfigException("Invalid EMR configuration: '" + node + "'");
+            }
+        }
+    }
+
+    private static class Filer
+    {
+        private final AmazonS3Client s3;
+        private final Optional<AmazonS3URI> staging;
+        private final Workspace workspace;
+        private final TemplateEngine templateEngine;
+        private final Config params;
+
+        private final List<StagingFile> files = new ArrayList<>();
```

Comment:
There's currently no deduplication implemented. It's just a collection of files. Changing this to Set would make it seem like there is a StagingFile identity and some kind of deduplication implemented even though there is none.

Perhaps deduplication would be a desirable feature, but I don't think it is necessary as staged files should not be large. I'm not sure the added complexity would be worth it.

 If it becomes necessary in the future, then we can implement it in a separate PR.